testing/integration/upgrade_test.go

```diff
@@ -0,0 +1,159 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+//go:build integration
+
+package integration
+
+import (
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/Masterminds/semver"
+	"github.com/google/uuid"
+
+	"github.com/elastic/elastic-agent-libs/kibana"
+
+	atesting "github.com/elastic/elastic-agent/pkg/testing"
+	"github.com/elastic/elastic-agent/pkg/testing/define"
+	"github.com/elastic/elastic-agent/pkg/testing/tools"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+)
+
+type UpgradeElasticAgent struct {
+	suite.Suite
+
+	requirementsInfo  *define.Info
+	agentStartVersion string
+	agentEndVersion   string
+	agentFixture      *atesting.Fixture
+}
+
+func newUpgradeElasticAgentTestSuite(info *define.Info, startVersion, endVersion string) *UpgradeElasticAgent {
+	return &UpgradeElasticAgent{
+		requirementsInfo:  info,
+		agentStartVersion: startVersion,
+		agentEndVersion:   endVersion,
+	}
+}
+
+// Before suite
+func (s *UpgradeElasticAgent) SetupSuite() {
+	agentFixture, err := atesting.NewFixture(
+		s.T(),
+		s.agentStartVersion,
+		atesting.WithFetcher(atesting.ArtifactFetcher()),
+	)
+	require.NoError(s.T(), err)
+	s.agentFixture = agentFixture
+}
+
+func (s *UpgradeElasticAgent) TestUpgradeFleetManagedElasticAgent() {
+	policyUUID := uuid.New().String()
+
+	createPolicyReq := kibana.CreatePolicyRequest{
+		Name:        "test-policy-" + policyUUID,
+		Namespace:   "default",
+		Description: "Test policy " + policyUUID,
+		MonitoringEnabled: []kibana.MonitoringEnabledOption{
+			kibana.MonitoringEnabledLogs,
+			kibana.MonitoringEnabledMetrics,
+		},
+	}
+	policy, err := s.requirementsInfo.KibanaClient.CreatePolicy(createPolicyReq)
+	require.NoError(s.T(), err)
+
+	createEnrollmentApiKeyReq := kibana.CreateEnrollmentAPIKeyRequest{
+		PolicyID: policy.ID,
+	}
+	enrollmentToken, err := s.requirementsInfo.KibanaClient.CreateEnrollmentAPIKey(createEnrollmentApiKeyReq)
+	require.NoError(s.T(), err)
+
+	// Get default fleet server URL
+	fleetServerURL, err := tools.GetDefaultFleetServerURL(s.requirementsInfo.KibanaClient)
+	require.NoError(s.T(), err)
+
+	output, err := tools.EnrollElasticAgent(fleetServerURL, enrollmentToken.APIKey, s.agentFixture)
+	if err != nil {
+		s.T().Log(string(output))
+	}
+	require.NoError(s.T(), err)
+
+	require.Eventually(s.T(), agentStatus("online", *s), 2*time.Minute, 10*time.Second, "Agent status is not online")
+
+	err = tools.UpgradeAgent(s.requirementsInfo.KibanaClient, s.agentEndVersion)
+	require.NoError(s.T(), err)
+
+	require.Eventually(s.T(), agentStatus("online", *s), 5*time.Minute, 5*time.Second, "Agent status is not online")
+
```

Comment:
TODO: we should wait 10 minutes here for the Upgrade Watcher to do it's thing and either successfully complete the upgrade or roll it back.